include/server/mir/scene/surface.h

```diff
@@ -136,6 +137,9 @@ class Surface :
     virtual auto application_id() const -> std::string = 0;
     virtual void set_application_id(std::string const& application_id) = 0;
     ///@}
+
+    /// The session this surface was created by
+    virtual auto session() const -> std::weak_ptr<Session> = 0;
```

Comment:
I'm concerned how users of this API can be proven to handle cases where the session is EOL but the surface owned by the session remains.

AFAICS that mapping of "surface" to "owning session" should only be needed in a small part of the mirserver implementation. Making it part of the mirserver API seems excessive.

Reply:
I think it's unclear how many places can benefit from this API because it doesn't exist yet. There are loads of places that store and pass around both a surface and a session (such as the entire `modify_window()` pipeline). In a future PR, I hope to switch them to only passing around a surface, and retrieving the session where necessary.

There's no single way to "prove" all users of the API correctly handle when the session dies before the surface, but individual cases can be tested.

Reply:
The thought occurs that another way to more directly skin this cat would be to remove the `Session` parameter from the `modify_window()` pipeline rather than try to make the `Session` accessible where people want to call `modify_window()`.

From my *very brief* skim this doesn't seem like it'd be too much churn. Particularly, `AbstractShell::modify_surface()` only uses the `session` to pass through to `WindowManager::modify_surface()`, and our primary user of this (`miral::WindowManager`) ignores the `session` parameter already.

Reply:
My plan *is* to remove session from `modify_surface()`, but I still think we should put this property here. For one thing, `AbstractShell::modify_surface()` calls `session->configure_streams()`. Furthermore, there are other places that currently hold both surfaces and sessions (such as `miral::Window::Self`). With this property those could also be simplified. Of could one could hold maps from surface -> session all over the place, but keeping the property on the surface makes more sense to me.

Reply:
I don't think anyone disputes that change is necessary, but it isn't clear that *this* change is the way to go.

To take your example, I see nothing in `ApplicationSession::configure_streams()` to explain why it needs [to be implemented on] session.

Reply:
Let's chat about this in our sync

Reply:
I believe the result of that sync was that while we may slowly move to architecture that doesn't need this property, adding it is the best step forward right now.